### PR TITLE
[css-tables] Add tests for visibility: collapse, visibility:hidden

### PR DESCRIPTION
**Pull request by joysyu**

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241274160)

# Firefox (nightly) #
Testing web-platform-tests at revision ed9dc414128617653dcfcac55b58989cfad4540b
Using browser at version BuildID 20170609100251; SourceStamp f4262773c4331d4ae139be536ce278ea9aad3436
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241274161)

# Sauce (safari) #
Testing web-platform-tests at revision ed9dc414128617653dcfcac55b58989cfad4540b
Using browser at version 10.0
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241274169)

# Chrome (unstable) #
Testing web-platform-tests at revision ed9dc414128617653dcfcac55b58989cfad4540b
Using browser at version 60.0.3112.20 dev
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241274170)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision ed9dc414128617653dcfcac55b58989cfad4540b
Using browser at version 14.14393
Starting 10 test iterations
No tests run.

**Comment by joysyu** [comment]
Looks like the timeout isn't related to my tests, but I'm not sure. Can someone confirm? Thanks!

**Comment by sideshowbarker** [comment]
w3c-test:mirror

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-col-002.html [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                <span style="height:50px"></span>
+            </x-td>
+        </x-tr>
+      </x-tbody>
+    </x-table>
+
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-col-002.html]: Please do not use `generate_tests()` - we are actively discouraging it. I know that we have throughout the testcases that are currently in WIP and you probably got that pattern from us. To fix this, use a for loop and then `assert_equals()`. You can see more on this on the [WPT Documentation](http://web-platform-tests.org/writing-tests/testharness.html).

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-col-003.html [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            <x-td>
+                <span style="height:50px"></span>
+            </x-td>
+        </x-tr>
+      </x-tbody>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-col-003.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-non-rowcol-001.html [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-non-rowcol-001.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-row-002.html [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-row-002.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-row-003.html [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-row-003.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-hidden-col-001.html [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            <x-td style="visibility:hidden;">
+                <span></span>
+            </x-td>
+        </x-tr>
+      </x-tbody>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-hidden-col-001.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-hidden-nested-001.html [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            <x-td>
+                <span id="invisibleSpan" style="visibility:hidden;"></span>
+            </x-td>
+        </x-tr>
+      </x-tbody>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-hidden-nested-001.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-hidden-nested-002.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                <span></span>
+            </x-td>
+        </x-tr>
+      </x-tbody>
+    </x-table>
+
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-hidden-nested-002.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-hidden-row-001.html [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-hidden-row-001.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-hidden-row-002.html [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    generate_tests(assert_equals, [
```
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-hidden-row-002.html]: same comment as above

File: css/work-in-progress/microsoft/css-tables/visibility-collapse-col-002.html (hunk 1 above)
Comment (gregwhitworth) [inline css/work-in-progress/microsoft/css-tables/visibility-collapse-col-002.html]: In addition, I know that I initially spoke with @davidsgrogan and we planned to put them in WIP, and he had the good point of just putting them into CSS3 tables path. We should begin on doing this, the path is `css/css-tables-3`. I won't block on this as @FremyCompany or myself will move them a bit later. I also think it's imperative that we resolve on the spec language for the sizing and painting of `visibility: collapse` within [issue 478](https://github.com/w3c/csswg-drafts/issues/478). @FremyCompany and myself will try to land this, so if you could please stage these in the actual tables folder that would be greatly appreciated, but not required for my sign off. Thanks!

**Review by gregwhitworth — CHANGES_REQUESTED** [review]
Overall it looks good @joysu. One thing that I'm not seeing is any testing of replaced elements nor dynamic modifications. Knowing that Edge currently has a bug in invalidation it would be useful to ensure that we're mimicking a user invalidation of some type. Thanks.

**Comment by joysyu** [comment]
Thanks so much @gregwhitworth for the code review! In addition to moving the tests to css-tables-3 and taking our generate_tests, I added three row span tests. I also added three tests with dynamic changes: visibility-collapse-col-004.html, visibility-collapse-row-004.html, and visibility-collapse-row-005.html (this also collapses an image). I'm not too sure what you mean by replaced elements; does collapsing an image count?

**Comment by joysyu** [comment]
Also, visibility-collapse-nested-001.html addresses the Edge bug :)

File: css/css-tables-3/visibility-collapse-row-005.html [hunk 11]
```diff
@@ -0,0 +1,89 @@
+<!doctype html>
```
Comment (gregwhitworth) [inline css/css-tables-3/visibility-collapse-row-005.html]: add dynamic to the end of this file name.

File: css/css-tables-3/visibility-collapse-row-005.html [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+            </x-td>
+            <x-td>
+                <span></span>
+            </x-td>
+        </x-tr>
+    </x-table>
+</main>
+
+<script>
+    colgroup = document.getElementById("collapse");
```
Comment (gregwhitworth) [inline css/css-tables-3/visibility-collapse-row-005.html]: you are getting a row, not a colgroup - change the variable name accordingly.

File: css/css-tables-3/visibility-collapse-rowspan-002.html [hunk 13]
```diff
@@ -0,0 +1,105 @@
+<!doctype html>
```
Comment (gregwhitworth) [inline css/css-tables-3/visibility-collapse-rowspan-002.html]: Can you please create a copy of `rowspan-002` & `rowspan-003` and create ones that are in `border-collapse: seperate` mode? Make sure this is reflected in the name so it's easy to differentiate the tests.

**Review by gregwhitworth — CHANGES_REQUESTED** [review]
@joysyu Overall this is really great. We're almost there, a few more changes as I tested them in FF/Edge and cross compared with the spec. Thank you so much - great work!

File: css/css-tables-3/visibility-collapse-rowspan-002.html (hunk 13 above)
Comment (joysyu) [inline css/css-tables-3/visibility-collapse-rowspan-002.html]: Done

File: css/css-tables-3/visibility-collapse-row-005.html (hunk 12 above)
Comment (joysyu) [inline css/css-tables-3/visibility-collapse-row-005.html]: Done

File: css/css-tables-3/visibility-collapse-row-005.html (hunk 11 above)
Comment (joysyu) [inline css/css-tables-3/visibility-collapse-row-005.html]: Done

File: css/css-tables-3/visibility-collapse-row-003.html [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+<script>
+    tests = [
+    [
+        document.getElementById('two').offsetWidth,
+        document.getElementById('one').offsetWidth,
+        "row group visibility:collapse doesn't change table width"
+    ],
+    [
+        document.getElementById('two').offsetHeight,
+        116,
```
Comment (joysyu) [inline css/css-tables-3/visibility-collapse-row-003.html]: Sorry this is actually the exact same test as visibility-collapse-nested-001, so I've deleted that and fixed this test to be more complex to catch Edge's bug. Thanks for catching this!

File: css/css-tables-3/visibility-collapse-nested-001.html [hunk 15]
```diff
@@ -0,0 +1,110 @@
+<!doctype html>
+<meta charset="utf-8">
+<script src='/resources/testharness.js'></script>
+<script src='/resources/testharnessreport.js'></script>
+<link rel='stylesheet' href='../support/base.css' />
+<link rel="author" title="Joy Yu" href="mailto:joysyu@mit.edu">
+<link rel="help" href="https://drafts.csswg.org/css-x-tables-3/#computing-the-x-table-height">
```
Comment (joysyu) [inline css/css-tables-3/visibility-collapse-nested-001.html]: Done

**Comment by joysyu** [comment]
Thank you @gregwhitworth! I've made the changes :) Let me know what you think!

**Comment by gsnedders** [comment]
@gregwhitworth @FremyCompany ping?

**Comment by davidsgrogan** [comment]
Joy eventually landed these through chromium's sync. This can be closed.

**Pull request closed without merging**